### ISSUE TITLE
Upgrade to OWLAPI 4

### ISSUE THREAD

**Issue by jamesaoverton** [ontodev/robot#12]
Better to do this sooner than later. Anyone have any experience with updating from 3 to 4?

We need to make sure the ELK and HermiT are happy with the OWLAPI 4.

**Comment by cmungall** [comment]
@hietze is this less painful than it used to be? They removed a lot of convenience methods.

**Comment by jamesaoverton** [comment]
I've been working on this. I had to implement some convenience methods. Everything seemed to work except that my filter operation is failing to copy multiple annotations on the same annotation axiom. I can see some related weirdness in Protege 5 but I'm having trouble constructing a minimal test case. I'll keep working on it.

**Comment by jamesaoverton** [comment]
I isolated the problem and filed an issue with OWLAPI: https://github.com/owlcs/owlapi/issues/406

It's a bigger bug than I thought. Under OWLAPI 4.0.2, all but one of the annotations on a given annotation are being silently dropped. This could cause serious damage to Uberon, for instance.

Until the bug is fixed, I think we should stick with OWLAPI 3.5.2.

**Comment by hdietze** [comment]
Actually 3.5.1 is the recommended version for the OWL-API (at the moment). In 3.5.2, There is a bug in the OWL2OBO translation: It does not fully remove IRIs and classes for ALT_IDs, this clutters up the OBO file obsolete Terms.

**Comment by hdietze** [comment]
Additional issues while switching to OWL-API 4.x:

* Reasoners are currently implemented against older OWL-API versions:
   * ELK: 3.4.5
   * HermiT: 3.4.10
* Performance: During the initial test some of the replacement methods seems to be very very slow and it was more efficient to iterate of the set of potential relevant axioms (i.e. axiom type) and filter on your own.

**Comment by jamesaoverton** [comment]
Ok, thanks. I'll downgrade from 3.5.2, and I'll tune the convenience methods (currently too simple).

**Comment by hdietze** [comment]
For the performance is was talking pure OWL-API methods, not the ROBOT code. They re-implemented some methods, which led for some of these to a performance (i.e. execution time) problem (for larger ontologies).

**Comment by ignazio1977** [comment]
Regarding performance, there was an issue in 4.0.1 with filters and another independent issue with Trove collections (we saved memory at the cost of speed). Both have been addressed in 4.0.2 - I think. If you have examples that are still very slow, I'd love to see them so I can profile.

Thanks for the test case in the bug report, very clear.

**Comment by hdietze** [comment]
Wonderful, the performance issue seems to be solved then. Correct, the initial tests with owltools were done using 4.0.1.

**Comment by sesuncedu** [comment]
ELK seems to be  working under 4.0.2 with just a few trivial changes (e.g. providing stub implementations for visiting annotation properties).

The time-performance issues in 4.0.1 were due to a combination of the time taken to create various axiom indices and the time taken to not use them :)

**Comment by jamesaoverton** [comment]
https://github.com/owlcs/owlapi/issues/406 has been closed as fixed, so we can try again when OWLAPI 4.0.3 is released. Thanks @ignazio1977!

@sesuncedu: Good to hear that ELK works with OWLAPI 4. I'd prefer to use an official ELK release, without adding stubs.

Anybody know whether HermiT works properly with OWLAPI 4? Thanks.

**Comment by ignazio1977** [comment]
On 11 May 2015 16:06, "James A. Overton" <notifications@github.com> wrote:
>
> owlcs/owlapi#406 has been closed as fixed, so we can try again when
OWLAPI 4.0.3 is released. Thanks @ignazio1977!
>

Will be 4.1.0 as we've decided to switch minor version for the next set of
changes. Not sure when though.

> @sesuncedu: Good to hear that ELK works with OWLAPI 4. I'd prefer to use
an official ELK release, without adding stubs.
>
> Anybody know whether HermiT works properly with OWLAPI 4? Thanks.
>

HermiT requires a few minor tweaks to work. I've got an experimental build
available on my github account.
The HermiT folks are aware of the changes and interested but they haven't
managed to merge them yet.

**Comment by hdietze** [comment]
OWL-API 4.1.4 seems to work for us. Aside from logging issues (OBO converter, trix parser) it seems to work fine.
ELK and Hermit are now both available for v4, so this is no longer blocked:
```
<dependency>
  <groupId>net.sourceforge.owlapi</groupId>
  <artifactId>org.semanticweb.hermit</artifactId>
  <version>1.3.8.413</version>
</dependency>
<dependency>
  <groupId>org.semanticweb.elk</groupId>
  <artifactId>elk-owlapi</artifactId>
  <version>0.4.3</version>
</dependency>

**Comment by jamesaoverton** [comment]
Closed by commit 5a09688d9c2abc1c157f86500086fba79471f6dc
